pkg/collector/corechecks/system/winregistry/winregistry.go

```diff
@@ -0,0 +1,210 @@
+	Subkeys map[string]subkeyCfg `yaml:"subkeys"` // The map key is the subkey name
+}
+
+type checkCfg struct {
+	RegistryKeys map[string]registryKeyCfg `yaml:"registry_keys"`
+}
+
+// registryKey is the in-memory representation of the key to monitor
+// it's different from the registryKeyCfg because we need to split the hive
+// from the keypath. It's easier to do this once during the Configure phase,
+// than every time the check runs.
+type registryKey struct {
+	name            string
+	hive            registry.Key
+	keyPath         string
+	originalKeyPath string // keep the original keypath around, for logging errors
+	subkeys         map[string]subkeyCfg
+}
+
+// WindowsRegistryCheck contains the field for the WindowsRegistryCheck
+type WindowsRegistryCheck struct {
+	core.CheckBase
+	metrics.Gauge
+	registryKeys []registryKey
+}
+
+// Configure reads the config and setups the check
+func (c *WindowsRegistryCheck) Configure(senderManager sender.SenderManager, integrationConfigDigest uint64, data integration.Data, initConfig integration.Data, source string) error {
+	err := c.CommonConfigure(senderManager, integrationConfigDigest, initConfig, data, source)
+	if err != nil {
+		return err
+	}
+
+	var conf checkCfg
+	if err := yaml.Unmarshal(data, &conf); err != nil {
+		return err
+	}
+
+	hiveMap := map[string]registry.Key{
+		"HKLM":               registry.LOCAL_MACHINE,
+		"HKEY_LOCAL_MACHINE": registry.LOCAL_MACHINE,
+		"HKU":                registry.USERS,
+		"HKEY_USERS":         registry.USERS,
+		"HKEY_CLASSES_ROOT":  registry.CLASSES_ROOT,
+		"HKCR":               registry.CLASSES_ROOT,
+	}
+
+	for regKey, regKeyConfig := range conf.RegistryKeys {
+		splitKeypath := strings.Split(regKey, "\\")
+		if len(splitKeypath) < 2 {
+			return log.Errorf("the key %s is too short to be a valid key", regKey)
+		}
+
+		if len(regKeyConfig.Name) == 0 {
+			log.Warnf("the key %s does not have a metric name, skipping", regKey)
+			continue
+		}
+
+		if hive, found := hiveMap[splitKeypath[0]]; found {
+			subkeys := make(map[string]subkeyCfg)
+			for valueName, regKeyCfg := range regKeyConfig.Subkeys {
+				if len(regKeyCfg.Name) == 0 {
+					log.Warnf("the subkey %s of %s does not have a metric name, skipping", valueName, regKey)
+				} else {
+					subkeys[valueName] = regKeyCfg
+				}
+			}
+			c.registryKeys = append(c.registryKeys, registryKey{
+				name:            regKeyConfig.Name,
+				hive:            hive,
+				originalKeyPath: regKey,
+				keyPath:         strings.Join(splitKeypath[1:], "\\"),
+				subkeys:         subkeys,
+			})
+		} else {
+			return log.Errorf("unknown hive %s", splitKeypath[0])
+		}
+	}
+
+	return nil
+}
+
+// Run runs the check
+func (c *WindowsRegistryCheck) Run() error {
+	sender, err := c.GetSender()
+	if err != nil {
+		return err
+	}
+
+	for _, regKeyCfg := range c.registryKeys {
+		regKey, err := registry.OpenKey(regKeyCfg.hive, regKeyCfg.keyPath, registry.QUERY_VALUE)
```

Comment:
Yes it would increase the key count. I think the limit is huge. But that interesting point. Both for your second question, I think. Also if one needs to open a key, write a value and close the key, proportionally first one will be slowest I believe. On the other hand in absolute term it should be small even for 100 or 1000 keys. Would be iteresting to know saving with one approach or another. There are a handful of other pros and cons. I think for simplicity sake it is quite adequate and not worth to discuss it further unless we would need it to optimize. Closing the comment.

Reply:
Hmm, interesting point. I wonder if it is true.

Reply:
Tested. It errors with 1018 (ERROR_KEY_DELETED)

Reply:
Feel free to resolve it for this PR. Something to keep in mind if we ever consider it in future.